src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +24,181 @@
-      cls.get_destination(),
+      in_dir,
       'logs',
       'exceptions{}.log'.format(intermediate_filename_component))
 
+  @classmethod
+  def log_exception(cls, msg):
+    """Try to log an error message to this process's error log and the shared error log.
+
+    NB: Doesn't raise (logs an error instead).
+    """
+    pid = os.getpid()
+    fatal_error_log_entry = cls._format_exception_message(msg, pid)
+
+    # We care more about this log than the shared log, so write to it first.
+    try:
+      cls._try_write_with_flush(cls._pid_specific_error_fileobj, fatal_error_log_entry)
+    except Exception as e:
+      logger.error(
+        "Error logging the message '{}' to the pid-specific file handle for {} at pid {}:\n{}"
+        .format(msg, cls._log_dir, pid, e))
+
+    # Write to the shared log.
+    try:
+      # TODO: we should probably guard this against concurrent modification by other pants
+      # subprocesses somehow.
+      cls._try_write_with_flush(cls._shared_error_fileobj, fatal_error_log_entry)
+    except Exception as e:
+      logger.error(
+        "Error logging the message '{}' to the shared file handle for {} at pid {}:\n{}"
+        .format(msg, cls._log_dir, pid, e))
+
+  @classmethod
+  def _try_write_with_flush(cls, fileobj, payload):
+    """This method is here so that it can be patched to simulate write errors.
+
+    This is because mock can't patch primitive objects like file objects.
+    """
+    fileobj.write(payload)
+    fileobj.flush()
+
+  @classmethod
+  def _check_or_create_new_destination(cls, destination):
+    try:
+      safe_mkdir(destination)
+    except Exception as e:
+      raise cls.ExceptionSinkError(
+        "The provided exception sink path at '{}' is not writable or could not be created: {}."
+        .format(destination, str(e)),
+        e)
+
+  @classmethod
+  def _recapture_fatal_error_log_streams(cls, new_log_location):
+    # NB: We do not close old file descriptors! This is bounded by the number of times any method is
+    # called, which should be few and finite.
+    # We recapture both log streams each time.
+    pid = os.getpid()
+    pid_specific_log_path = cls.exceptions_log_path(for_pid=pid, in_dir=new_log_location)
+    shared_log_path = cls.exceptions_log_path(in_dir=new_log_location)
+    assert(pid_specific_log_path != shared_log_path)
+    try:
+      # Truncate the pid-specific error log file.
+      pid_specific_error_stream = safe_open(pid_specific_log_path, mode='w')
+      # Append to the shared error file.
+      shared_error_stream = safe_open(shared_log_path, mode='a')
+    except Exception as e:
+      raise cls.ExceptionSinkError(
+        "Error opening fatal error log streams for log location '{}': {}"
+        .format(new_log_location, str(e)))
+
+    # TODO: determine whether any further validation of the streams (try writing to them here?) is
+    # useful/necessary for faulthandler (it seems it just doesn't write to e.g. closed file
```

Comment:
If it is the case that `faulthandler` doesn't write to closed file descriptors, it seems like it should be safe to close them? But see above regarding calling `unregister`/`disable`.

Reply:
I think whatever python does in order to give you the fileobj to write to should be enough "validation", this is just a reminder to think about it. Will remove.

Reply:
Removed, replaced with a link to the faulthandler backport source to see what `enable()` and `disable()` do.